lib/SILOptimizer/Utils/Generics.cpp

```diff
@@ -385,6 +387,10 @@ static bool createsInfiniteSpecializationLoop(ApplySite Apply) {
                  Replacement->dump(llvm::dbgs());
                });
 
+    if (!visited.insert(CurSpecializationInfo).second) {
```

Comment:
I think you can replace both the growingSubstitutions() check and this with a termination count. It is undecidable if recursively applying a substitution map to itself will eventually cycle back, or continue producing an infinite set of distinct substitution maps.

Reply:
Yes, I think this needs to be re-designed anyway. For now I just want to fix the immediate problem.